compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -4101,9 +4098,34 @@ private STNode parseExpression(OperatorPrecedence precedenceLevel, boolean isRhs
     private STNode parseExpression(SyntaxKind kind, OperatorPrecedence precedenceLevel, boolean isRhsExpr,
                                    boolean allowActions, boolean isInMatchGuard, boolean isInConditionalExpr) {
         STNode expr = parseTerminalExpression(kind, isRhsExpr, allowActions, isInConditionalExpr);
-        return parseExpressionRhs(precedenceLevel, expr, isRhsExpr, allowActions, isInMatchGuard, isInConditionalExpr);
+        STNode actionOrExpression =
+                parseExpressionRhs(precedenceLevel, expr, isRhsExpr, allowActions, isInMatchGuard, isInConditionalExpr);
+        //braced actions are just paranthesis enclosing actions, no need to add a diagnostic there when we have added
+        //diagnostics to its children
+        if (!allowActions && isAction(actionOrExpression) && actionOrExpression.kind != SyntaxKind.BRACED_ACTION) {
+            actionOrExpression = attachErrorExpectedActionFoundDiagnostic(actionOrExpression);
+        }
+        return actionOrExpression;
     }
 
+    private boolean containsDiagnostics(STNode node, DiagnosticErrorCode diagnosticErrorCode) {
```

Comment:
Why do we need this check?
Can the same error be logged in two places? If so, Would removing the error from the other place solve the problem?

Reply:
no I dont think that will work, because there are loops and multiple entry points for the parseExpression method.

Reply:
I ll try to get it working like that